### ISSUE TITLE
HELP: Improve documentation

### ISSUE THREAD

**Issue by aolofsson** [parallella/pal#3]
Task:
-Improve documentation

Guideliens:
-documentation at the beginning of each function source file
-descriptions should be as short as possible but not shorter
-doxygen compatible
-explain basics and anything unusual about function

For:
-anyone

**Comment by indrora** [comment]
Is there better documentation on what needs to be done in terms of device-specificness? How code is managed across several different devices? teams?

Namely,

* will teams always be comprised of homogenized devices (that is, a team will contain only one kind of device)
* When writing this code, should we be writing it assuming the code we're writing is going to be running on the target system (pal will be compiled into the code running into the target system (i.e. epiphany, GPU) executable) or being run by a host of some kind (e.g. an ARM processor coordinating several Epiphany chips)?

**Comment by tcptomato** [comment]
Or better said. A description on the vision on this library. How, where will it be used. At the moment my impression is that's mostly target at epiphany.

**Comment by indrora** [comment]
Yeah. There's so many things that I want to implement (namely, mutexes) that I would like to make sure are done homogenously with the rest of the codebase, but there's no good set of rules on what that looks like.

Worse, there's no good examples in the source base on what a specific device looks like (except for the demo device which isn't well defined.) Is there a copy of Doxygen documentation that's been run over the codebase as it stands?
